test/main_dialog_spec.rb

```diff
@@ -0,0 +1,174 @@
+      mock_ui_events(:edit_alternative, :cancel)
+      allow(Yast::UI).to receive(:QueryWidget).with(:alternatives_table, :CurrentItem).and_return(2)
+    end
+
+    it "opens an AlternativeDialog with the selected alternative" do
+      selected_alternative = dialog.instance_variable_get(:@alternatives_list)[2]
+      alternative_dialog = UpdateAlternatives::AlternativeDialog.new(selected_alternative)
+
+      expect(UpdateAlternatives::AlternativeDialog).to receive(:new)
+        .with(selected_alternative)
+        .and_return(alternative_dialog)
+      expect(alternative_dialog).to receive(:run)
+
+      dialog.run
+    end
+
+    it "updates the alternatives table" do
+      alternative_dialog = UpdateAlternatives::AlternativeDialog.new(nil)
+      allow(UpdateAlternatives::AlternativeDialog).to receive(:new)
+        .and_return(alternative_dialog)
+      allow(alternative_dialog).to receive(:run)
+
+      expect(Yast::UI).to receive(:ChangeWidget).with(:alternatives_table, :Items, kind_of(Array))
+
+      dialog.run
+    end
+  end
+
+  describe "#search_handler" do
+    before do
+      # First we send :multi_choice_only event to disable the filter.
+      mock_ui_events(:multi_choice_only, :search, :cancel)
+      allow(Yast::UI).to receive(:QueryWidget).with(:multi_choice_only, :Value).and_return(false)
+      # Expect fill the alternatives table when opening the dialog.
+      expect_update_table_with(
+        [
+          [Id(0), "editor", "/usr/bin/vim", "auto"],
+          [Id(1), "pip", "/usr/bin/pip3.4", "auto"],
+          [Id(2), "test", "/usr/bin/test2", "manual"]
+        ]
+      )
+    end
+
+    context "if the input field is empty" do
+      it "shows all alternatives" do
+        allow(Yast::UI).to receive(:QueryWidget).with(:search, :Value).and_return("")
+        expect_update_table_with(
+          [
+            [Id(0), "editor", "/usr/bin/vim", "auto"],
+            [Id(1), "pip", "/usr/bin/pip3.4", "auto"],
+            [Id(2), "test", "/usr/bin/test2", "manual"]
+          ]
+        )
+        dialog.run
+      end
+    end
+
+    context "if the input field has text that match with some alternative's name" do
+      it "shows the alternatives who match its name with the text" do
+        allow(Yast::UI).to receive(:QueryWidget).with(:search, :Value).and_return("ed")
+        expect_update_table_with([[Id(0), "editor", "/usr/bin/vim", "auto"]])
+        dialog.run
+      end
+    end
+
+    context "if the input field has text that does not match with any alternative's name" do
+      it "does not show any alternative" do
+        allow(Yast::UI).to receive(:QueryWidget).with(:search, :Value).and_return("no match")
+        expect_update_table_with([])
+        dialog.run
+      end
+    end
+  end
+
+  describe "#accept_handler" do
+    before do
+      mock_ui_events(:accept)
+    end
+
+    it "saves all changes" do
+      expect(dialog.instance_variable_get(:@alternatives_list)).to all receive(:save)
```

Comment:
At first sight, it looks to me like this could be implemented without using `#instance_variable_get` (which is nasty) if you would use a `let` to assign/hold the alternatives list.